lisa/src/main/java/it/unive/lisa/analysis/nonrelational/Environment.java

```diff
@@ -116,29 +116,49 @@ public M assume(E expression, ProgramPoint pp) throws SemanticException {
 		else if (lattice.satisfies(expression, (M) this, pp) == Satisfiability.SATISFIED)
 			return (M) this;
 		else
-			return assumeAux(expression, pp);
+			return glb(lattice.assume((M) this, expression, pp));
 	}
 
 	/**
-	 * Auxiliary version of {@link #assume(SymbolicExpression, ProgramPoint)}
-	 * where the cases where the expression is never satisfied
-	 * ({@code lattice.satisfies(expression, this) == Satisfiability.NOT_SATISFIED})
-	 * and is always satisfied
-	 * ({@code lattice.satisfies(expression, this) == Satisfiability.SATISFIED})
-	 * have already been handled. The given expression thus holds sometimes.
+	 * Performs the greatest lower bound between this environment and
+	 * {@code other}.
 	 * 
-	 * @param expression the expression to assume to hold.
-	 * @param pp         the program point that where this operation is being
-	 *                       evaluated
+	 * @param other the other environment
 	 * 
-	 * @return the (optionally) modified copy of this domain
+	 * @return the greatest lower bound between this environment and
+	 *             {@code other}
 	 * 
-	 * @throws SemanticException if an error occurs during the computation
+	 * @throws SemanticException if something goes wrong during the computation
 	 */
-	protected M assumeAux(E expression, ProgramPoint pp) throws SemanticException {
-		// TODO: a more precise filtering is needed when satisfiability of
-		// expression is unknown - subclasses might add some logic
-		return copy();
+	@SuppressWarnings("unchecked")
+	public M glb(M other) throws SemanticException {
+		if (other == null || this.isBottom() || other.isTop() || this == other || this.equals(other)
+				|| this.lessOrEqual(other))
+			return (M) this;
+
+		if (other.isBottom() || this.isTop() || other.lessOrEqual((M) this))
+			return (M) other;
+
+		return glbFunctionalLift(other, (o1, o2) -> o1 == null ? o2 : o1.glb(o2));
+	}
+
+	private final M glbFunctionalLift(M other, FunctionalLift<T> lift) throws SemanticException {
```

Comment:
This is basically the same code as `lubFuncionalLift` from `FunctionalLattice`. We should go back to a generic `functionaLift` that simply takes an additional parameter that decides how to obtain the key set.